File: go/bank-account/bank_account.go
package account

import (
	"sync"
)

// Account struct to save bank transactions
type Account struct {
	mux    sync.Mutex
	Amount int64
	Closed bool
}

// Balance function to return the amount value
func (a *Account) Balance() (int64, bool) {
	if a.Amount >= 0 && !a.Closed {
		return a.Amount, true
	}

	return a.Amount, false
}

// Deposit function to add more money in account
func (a *Account) Deposit(amount int64) (int64, bool) {
	a.mux.Lock()
	defer a.mux.Unlock()

	if a.Closed {
		return a.Amount, false
	}

	if amount < 0 && a.Amount == 0 {
		return 0, false
	}

	a.Amount = a.Amount + amount

	return a.Amount, true
}

// Close function to close account and retrieve the amount value
func (a *Account) Close() (int64, bool) {
	var payout int64

	a.mux.Lock()
	defer a.mux.Unlock()

	if !a.Closed {
		payout = a.Amount
		a.Amount = 0
		a.Closed = true

		return payout, true
	}

	return payout, false
}

// Open function to open a new account
func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}

	var account Account = Account{
		Amount: amount,
		Closed: false,
	}

	return &account
}

File: python/bank-account/bank_account.py
from threading import Lock

class BankAccount:
    def __init__(self):
        self.amount = 0
        self.closed = True
        self.lock = Lock()

    def get_balance(self):
        if self.closed:
            raise ValueError("Your account was closed")

        return self.amount

    def open(self):
        if not self.closed:
            raise ValueError("You already have account")

        self.closed = False

    def deposit(self, amount):
        if self.closed:
            raise ValueError("Your account was closed")

        if amount < 0:
            raise ValueError("You can't deposite negative values")

        with self.lock:
            self.amount += amount

    def withdraw(self, amount):
        if self.closed:
            raise ValueError("Your account was closed")

        if amount < 0:
            raise ValueError("You can't deposit negative values")

        if self.amount < amount:
            raise ValueError("You can't withdraw more than your balance")

        with self.lock:
            self.amount -= amount

    def close(self):
        if self.closed:
            raise ValueError("You have already closed your account")            

        with self.lock:
            self.amount = 0
            self.closed = True
